pkg/webhook/webhook.go

```diff
@@ -175,3 +192,102 @@ func (c *Webhook) validateHandlerFunc(review *admission_v1beta1.AdmissionReview)
 		Allowed: true,
 	}
 }
+
+func (c *Webhook) validateRelease(request *admission_v1beta1.AdmissionRequest, release shipper.Release) error {
+	var err error
+	overrideRBs := rolloutblock.NewOverride(release.Annotations[shipper.RolloutBlocksOverrideAnnotation])
+	err = c.validateOverrideRolloutBlockAnnotation(overrideRBs, release.Namespace)
+	if err != nil {
+		return err
+	}
+
+	switch request.Operation {
+	case kubeclient.Create:
+		err = c.processRolloutBlocks(release.Namespace, overrideRBs)
+	case kubeclient.Update:
+		var oldRelease shipper.Release
+		err = json.Unmarshal(request.OldObject.Raw, &oldRelease)
+		if err != nil {
+			return nil
+		}
+		newOverrides := rolloutblock.NewOverride(release.Annotations[shipper.RolloutBlocksOverrideAnnotation])
+		oldOverrides := rolloutblock.NewOverride(oldRelease.Annotations[shipper.RolloutBlocksOverrideAnnotation])
+		subtractedOverrides := oldOverrides.Diff(newOverrides)
+		if len(subtractedOverrides) == 0 {
+			// no overrides were removed from annotation. validating release:
+			err = c.processRolloutBlocks(release.Namespace, overrideRBs)
+		} else {
+			// rolloutblock override were changed, making sure spec was not changed
+			if !reflect.DeepEqual(release.Spec, oldRelease.Spec) {
+				err = fmt.Errorf("releases Spec was changed %v", release)
+			}
+		}
+	default:
+	}
+
+	return err
+}
+
+func (c *Webhook) validateApplication(request *admission_v1beta1.AdmissionRequest, application shipper.Application) error {
+	var err error
+	overrideRBs := rolloutblock.NewOverride(application.Annotations[shipper.RolloutBlocksOverrideAnnotation])
+	err = c.validateOverrideRolloutBlockAnnotation(overrideRBs, application.Namespace)
+	if err != nil {
+		return err
+	}
+
+	if request.Operation == kubeclient.Create {
```

Comment:
I believe we should also validate against rollout blocks if we're trying to roll out an existing application. Currently, changing any of the values in the spec triggers rollout blocks, but the webhook does not return an error.

Reply:
So the Update case is a bit tricky, and even though webhook doesn't validate againts rollout block in updates, application and release controller will still maintain blocking mechanism. Webhook should accept the change in case the change it in annotation (like removing an override in order to block the object). If we validate against rollout blocks in case on an update as well, webhook will reject cases like these. I can add a check to see if it's a case of removed annotation, it might cause the validating function to be a bit weird looking but I think it will be just fine...

Reply:
I suppose the logic would be to check against rollout blocks if the spec (which is the only thing that triggers rollouts) was changed, and otherwise just let the changes through. That way, if one is changing labels, annotations or anything else, the webhook won't reject the update. This logic should apply to releases as well.